homeassistant/components/hive/__init__.py

```diff
@@ -52,101 +34,83 @@
-    }
-)
-
 
 async def async_setup(hass, config):
-    """Set up the Hive Component."""
-
-    async def heating_boost(service):
-        """Handle the service call."""
-
-        entity_lookup = hass.data[DOMAIN]["entity_lookup"]
-        hive_id = entity_lookup.get(service.data[ATTR_ENTITY_ID])
-        if not hive_id:
-            # log or raise error
-            _LOGGER.error("Cannot boost entity id entered")
-            return
-
-        minutes = service.data[ATTR_TIME_PERIOD]
-        temperature = service.data[ATTR_TEMPERATURE]
-
-        hive.heating.turn_boost_on(hive_id, minutes, temperature)
-
-    async def hot_water_boost(service):
-        """Handle the service call."""
-        entity_lookup = hass.data[DOMAIN]["entity_lookup"]
-        hive_id = entity_lookup.get(service.data[ATTR_ENTITY_ID])
-        if not hive_id:
-            # log or raise error
-            _LOGGER.error("Cannot boost entity id entered")
-            return
-        minutes = service.data[ATTR_TIME_PERIOD]
-        mode = service.data[ATTR_MODE]
+    """Set up the Hive Integration."""
+    hass.data[DOMAIN] = {"entries": {}}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    if not hass.config_entries.async_entries(DOMAIN):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: conf[CONF_USERNAME],
+                    CONF_PASSWORD: conf[CONF_PASSWORD],
+                },
+            )
+        )
+    return True
 
-        if mode == "on":
-            hive.hotwater.turn_boost_on(hive_id, minutes)
-        elif mode == "off":
-            hive.hotwater.turn_boost_off(hive_id)
 
-    hive = Hive()
+async def async_setup_entry(hass, entry):
+    """Set up Hive from a config entry."""
 
-    config = {}
-    config["username"] = config[DOMAIN][CONF_USERNAME]
-    config["password"] = config[DOMAIN][CONF_PASSWORD]
-    config["update_interval"] = config[DOMAIN][CONF_SCAN_INTERVAL]
+    websession = aiohttp_client.async_get_clientsession(hass)
+    hive = Hive(websession)
+    hive_config = dict(entry.data)
 
-    devices = await hive.session.startSession(config)
+    hive_config["options"] = {}
+    hive_config["options"].update(
+        {CONF_SCAN_INTERVAL: dict(entry.options).get(CONF_SCAN_INTERVAL, 120)}
+    )
+    hass.data[DOMAIN]["entries"][entry.entry_id] = hive
 
-    if devices is None:
-        _LOGGER.error("Hive API initialization failed")
-        return False
+    try:
+        devices = await hive.session.startSession(hive_config)
+    except HTTPException as error:
+        _LOGGER.error("Could not connect to the internet: %s", error)
+        raise ConfigEntryNotReady() from error
 
-    hass.data[DOMAIN][DATA_HIVE] = hive
-    hass.data[DOMAIN]["entity_lookup"] = {}
+    if devices == "INVALID_REAUTH":
```

Comment:
Side note: It's weird that `devices` both can be a string representing an error and a dict with hive types. I'd recommend instead to raise an exception when the authentication fails and we should reauthenticate. Return types should be as specific as possible.